### PR TITLE
Add missing Java and Fortran bindings

### PR DESCRIPTION
**Pull request by ggouaillardet**

### CODE REVIEW

**Comment by lanl-ompi** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.open-mpi.org/job/ompi_master_pr_cle5.2up02/89/
Test PASSed.

**Comment by mellanox-github** [comment]
* Coverity found 816 errors for all

* Coverity found 5 errors for oshmem

* Coverity found 2 errors for fca
* Coverity report at http://bgate.mellanox.com:8888/jenkins/job/gh-ompi-master-pr//587/Coverity_Report <br><br>:octocat: *Sent from [GH](http://nodegh.io).*

**Comment by mellanox-github** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://bgate.mellanox.com/jenkins/job/gh-ompi-master-pr/587/
Test PASSed.

**Comment by jsquyres** [comment]
@ggouaillardet Are you sure that these C++ bindings exist?  We deleted the C++ bindings in MPI-3.0 -- so it's quite possible that the C++ bindings for some of these new RMA functions don't actually exist...?

**Comment by ggouaillardet** [comment]
i will double check that (i was being symmetric, so i did not even check these bindings existence)

btw, i saw MPI 3.1 fixed f08 bindings.
does this preserve binary compatibility with MPI 3.0 ?
if not, does it mean MPI 3.1 f08 bindings will not be available in v1.8 and v1.10 ?

**Comment by lanl-ompi** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.open-mpi.org/job/ompi_master_pr_distcheck/39/
Test PASSed.

**Comment by goodell** [comment]
> @ggouaillardet Are you sure that these C++ bindings exist? We deleted the C++ bindings in MPI-3.0 -- so it's quite possible that the C++ bindings for some of these new RMA functions don't actually exist...?

To be clearer on this point: it's not OK to add bindings based on what seems like a natural adaptation of the C bindings.  Each binding which is added should be checked against the latest version of the MPI standard to make sure that we match other implementations and comply to the standard as much as possible.  If the binding in the standard seems wrong, then we should raise the issue with the MPI Forum first.

**Comment by lanl-ompi** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.open-mpi.org/job/ompi_master_pr_distcheck/42/
Test PASSed.

**Comment by lanl-ompi** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.open-mpi.org/job/ompi_master_pr_cle5.2up02/135/
Test PASSed.

**Comment by mellanox-github** [comment]
* Coverity found 812 errors for all

* Coverity found 5 errors for oshmem

* Coverity found 2 errors for fca
* Coverity report at http://bgate.mellanox.com:8888/jenkins/job/gh-ompi-master-pr//624/Coverity_Report <br><br>:octocat: *Sent from [GH](http://nodegh.io).*

**Comment by mellanox-github** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://bgate.mellanox.com/job/gh-ompi-master-pr/624/

**Comment by ggouaillardet** [comment]
since these C++ bindings are not in the standard, i did remove them
(and remove C++ refererences from the man pages)

@jsquyres could you please review the f08 part ?

File: ompi/mpi/fortran/use-mpi-f08/attr-fn-f08-callback-interfaces.h [hunk 1]
```diff
@@ -43,7 +45,7 @@ interface
 
   subroutine MPI_COMM_NULL_COPY_FN( comm, comm_keyval, extra_state, &
                                     attribute_val_in, attribute_val_out, &
-                                    flag, ierr )
+                                    flag, ierr ) BIND(C,name="ompi_comm_null_copy_fn_f")
```
Comment (jsquyres) [inline ompi/mpi/fortran/use-mpi-f08/attr-fn-f08-callback-interfaces.h]: Can you explain why these subroutines need to be BIND(C) to a specific C function name?

File: ompi/mpi/fortran/use-mpi-f08/mpi-f08-interfaces-callbacks.F90 [hunk 2]
```diff
@@ -64,7 +66,7 @@ SUBROUTINE MPI_User_function(invec, inoutvec, len, datatype)
 
 OMPI_ABSTRACT INTERFACE
 SUBROUTINE MPI_Comm_copy_attr_function(oldcomm,comm_keyval,extra_state, &
-                                       attribute_val_in,attribute_val_out,flag,ierror)
+                                       attribute_val_in,attribute_val_out,flag,ierror) BIND(C)
```
Comment (jsquyres) [inline ompi/mpi/fortran/use-mpi-f08/mpi-f08-interfaces-callbacks.F90]: Can you explain why these subroutines need to be BIND(C)?

File: ompi/mpi/man/man3/MPI_Win_attach.3in [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+.SH SYNTAX
+.ft R
+.SH C Syntax
+.nf
+#include <mpi.h>
+MPI_Win_attach(MPI_Win *\fIwin\fP, void *\fIbase\fP, MPI_Aint \fIsize\fP)
+
+MPI_Win_detach(MPI_Win *\fIwin\fP, void *\fIbase\fP)
+.fi
+.SH Fortran Syntax (see FORTRAN 77 NOTES)
```
Comment (jsquyres) [inline ompi/mpi/man/man3/MPI_Win_attach.3in]: Delete the "(see Fortran 77 notes)"

File: ompi/mpi/man/man3/MPI_Win_attach.3in [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+If the \fIbase\fP value used by MPI_Win_create was allocated by MPI_Alloc_mem, the size of the window can be no larger than the value set by the MPI_ALLOC_MEM function.
+.sp
+
+.SH NOTES
+Use memory allocated by MPI_Alloc_mem to guarantee properly aligned window boundaries (such as word, double-word, cache line, page frame, and so on).
+.sp
+ 
+
+
+.SH FORTRAN 77 NOTES
```
Comment (jsquyres) [inline ompi/mpi/man/man3/MPI_Win_attach.3in]: Delete this Fortran 77 notes section.

File: ompi/mpi/man/man3/MPI_Win_create_dynamic.3in [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+.SH SYNTAX
+.ft R
+.SH C Syntax
+.nf
+#include <mpi.h>
+MPI_Win_create_dynamic(MPI_Info \fIinfo\fP, MPI_Comm \fIcomm\fP, MPI_Win *\fIwin\fP)
+
+.fi
+.SH Fortran Syntax (see FORTRAN 77 NOTES)
```
Comment (jsquyres) [inline ompi/mpi/man/man3/MPI_Win_create_dynamic.3in]: Delete the "(see Fortran 77 notes)"

File: ompi/mpi/man/man3/MPI_Win_create_dynamic.3in [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+.sp 
+.TP 1i
+accumulate_ops
+If set to \fIsame_op\fP, the implementation will assume that all concurrent
+accumulate calls to the same target address will use the same
+operation.  If set to \fIsame_op_no_op\fP, then the implementation will
+assume that all concurrent accumulate calls to the same target address
+will use the same operation or MPI_NO_OP.  The default is \fIsame_op_no_op\fP.
+
+.SH FORTRAN 77 NOTES
```
Comment (jsquyres) [inline ompi/mpi/man/man3/MPI_Win_create_dynamic.3in]: Delete this Fortran 77 Notes section.

**Comment by jsquyres** [comment]
Should this be targeted at v1.10.0 instead of 1.8.x?  (sorry; it was probably my delay that caused this to miss the v1.8.6 release :frowning: )

**Comment by ggouaillardet** [comment]
these subroutines are used as parameters of other subroutines.
I tried more "natural" things, but test cases failed to compile :-(
I could probably define them as subroutine calling the C function, but I do not see any value in doing this.
any suggestion ?

**Comment by hppritcha** [comment]
@nrgraham23  may want to follow this discussion since he's working on java bindings as well.

**Comment by jsquyres** [comment]
Wait, this conversation is getting mixed between Java and Fortran bindings.  @ggouaillardet is replying to my question about his use of `BIND(C)` in the Fortran bindings.

Ok, I think I see why they need to be `BIND(C)`.  If you can delete those sections from the man pages, I think this PR is good to go.  :+1:

**Comment by jsquyres** [comment]
Oh, one last comment: can you add "Fixes #460" in the commit message for the commit that adds the missing fortran functions?

Thanks!

**Comment by ggouaillardet** [comment]
done !

**Comment by mellanox-github** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://bgate.mellanox.com/job/gh-ompi-master-pr/674/

**Pull request merged**